From 137a754a59c6fb537d3c78bf1a04e07b7c21f279 Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Wed, 5 Jul 2023 10:50:36 -0400
Subject: [PATCH 32/55] Detect Moto dynamic hardware feature

Moto added a custom node in sysconfig XMLs:
<unavailable-feature-conditional />
This node reads a property and enables a feature based on it.

Take those into account to enable NFC on Moto devices which have
NFC-less variants
---
 .../java/com/android/server/SystemConfig.java | 24 +++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/services/core/java/com/android/server/SystemConfig.java b/services/core/java/com/android/server/SystemConfig.java
index 2dca73ff7..c09824c87 100644
--- a/services/core/java/com/android/server/SystemConfig.java
+++ b/services/core/java/com/android/server/SystemConfig.java
@@ -926,6 +926,9 @@ public class SystemConfig {
                     case "unavailable-feature":
                         readUnavailableFeature(parser, permFile, allowFeatures, name);
                         break;
+                    case "unavailable-feature-conditional":
+                        readUnavailableFeatureConditional(parser, permFile, allowFeatures, name);
+                        break;
                     case "allow-in-power-save-except-idle":
                         readAllowInPowerSaveExceptIdle(
                                 parser, permFile, allowOverrideAppRestrictions, name);
@@ -1265,6 +1268,27 @@ public class SystemConfig {
         XmlUtils.skipCurrentTag(parser);
     }

+    private void readUnavailableFeatureConditional(XmlPullParser parser, File permFile,
+            boolean allowFeatures, String name) throws IOException, XmlPullParserException {
+        if (allowFeatures) {
+            String fname = parser.getAttributeValue(null, "name");
+            String prop = parser.getAttributeValue(null, "prop");
+            if (fname == null || prop == null) {
+                Slog.w(TAG, "<" + name + "> without name or prop in " + permFile
+                        + " at " + parser.getPositionDescription());
+            } else {
+                if (SystemProperties.getBoolean(prop, false)) {
+                    addFeature(fname, 0);
+                } else {
+                    mUnavailableFeatures.add(fname);
+                }
+            }
+        } else {
+            logNotAllowedInPartition(name, permFile, parser);
+        }
+        XmlUtils.skipCurrentTag(parser);
+    }
+
     private void readAllowInPowerSaveExceptIdle(XmlPullParser parser, File permFile,
             boolean allow, String name) throws IOException, XmlPullParserException {
         if (allow) {
--
2.43.0
